### PR TITLE
Karpathy's AutoResearch with Memory-in-the-Loop States by The Adimension

### PR DESCRIPTION
**Pull request by habanwer**
## Summary

This fork applies the [DEITY Principles Framework](https://doi.org/10.1093/ehjimp/qyaf038) (Data, Ethics, Informatics, Technology, You) to restructure autoresearch for transparent human-machine collaboration across GPU platforms — from Volta (SM 7.0) through Blackwell (SM 10.0).

## Changes (8 per-file commits)

| File | Type | Description |
|---|---|---|
| `ground.json` | NEW | User-owned read-only config: data paths, tokenizer, time budgets, processor overrides |
| `model.json` | NEW | Agent-owned hyperparameters: architecture, optimization, evaluation |
| `prepare.py` | MODIFIED +125/-17 | GPU platform detection, ground.json reads, `PLATFORM` dict export |
| `train.py` | MODIFIED +194/-55 | model.json reads, PLATFORM import, fp32 moments, crash handler, `update_research_memory()` |
| `program.md` | MODIFIED +96/-114 | Structured agent protocol with file ownership governance |
| `.gitignore` | MODIFIED +5/-2 | Add *.pkl, run.log; remove results.tsv from ignore |
| `analysis.ipynb` | MODIFIED metadata | Kernel updated to Python 3.12.10 |
| `README.md` | MODIFIED | Fork introduction with DEITY Principles mapping |

## Key design decisions

- **Config extraction**: Hardcoded constants → `ground.json` (user-owned) + `model.json` (agent-owned)
- **File ownership governance**: Clear boundaries for human/agent through program.md
- **GPU auto-detection**: dtype, attention backend, torch.compile, GradScaler per GPU generation
- **Memory-in-the-loop**: `update_research_memory()` persists outcomes to `sessions/memory.md` so the agent's next hypothesis is informed by all prior runs, in addition to logs, and standard structure of tab-separated values.
- **Windows support**: compile guards for sys.platform, extended in reference to notable repo: [jsegov/autoresearch-win-rtx](https://github.com/jsegov/autoresearch-win-rtx)

## Attribution
- Upstream: [karpathy/autoresearch](https://github.com/karpathy/autoresearch)
- Platform support: [jsegov/autoresearch-win-rtx](https://github.com/jsegov/autoresearch-win-rtx)
- Framework: [The Adimension: DEITY Principles](https://doi.org/10.1093/ehjimp/qyaf038) — *Eur Heart J Imaging Methods Pract*, Shehab Anwer, 2025.

Author: Shehab Anwer, MD — [habanwer](https://github.com/habanwer) · [The Adimension](https://github.com/the-adimension)

### CODE REVIEW

**Comment by habanwer** [comment]
@copilot open a new pull request to apply changes based on the comments in [this thread](https://github.com/karpathy/autoresearch/pull/302#pullrequestreview-3955893850)

**Comment by svlandeg** [comment]
> @copilot open a new pull request to apply changes based on the comments in [this thread](https://github.com/karpathy/autoresearch/pull/302#pullrequestreview-3955893850)

Hi @habanwer! 

My two cents: it's much nicer if you would iterate on the PR first locally, including asking an LLM to review and making corresponding edits etc, all BEFORE you open the PR to merge to upstream. 

If you're asking copilot to review on the PR online, you'll be expected by OS maintainers to also at least address all those review comments (comment on them, apply them, resolve them). Opening multiple PRs for the same issue/feature adds a lot of noise/notications for maintainers.

**Comment by habanwer** [comment]
Dear @svlandeg Thanks for your response, I am terribly sorry I clicked it by mistake, and then it ran - just noticed that! I will try to revert that or resubmit the pull - thanks again!

**Comment by habanwer** [comment]
Dear @svlandeg 

Thank you for the guidance — I've taken it to heart. The `@copilot` trigger was genuinely accidental, and I apologize for the noise. I managed it locally and then force-pushed a single commit (6181369) addressing the substantive findings from the automated review:

**train.py:**
- Removed unused `import re`
- Added dtype cast in `apply_rotary_emb` so rotary embeddings match the model's autocast dtype instead of forcing fp32 upcast
- Guarded `eval_steps` with `max(1, ...)` to prevent *division-by-zero* crash specifically when token budget is small
- Replaced unconditional `torch.compile` disable with `PLATFORM["compile"]` so Linux/Triton platforms benefit from compilation
- Added clarifying comment on SDPA boolean mask convention (verified `True = allowed` against PyTorch 2.6 — the bot's claim of inverted semantics was inaccurate - for your kind review and approval)

**prepare.py:**
- Replaced bare `open()` with explicit error handling (`FileNotFoundError`, `JSONDecodeError`) and top-level key validation — comment updated to *reflect `ground.json` is required*, not optional.

**program.md:**
- Fixed `results.tsv` header to show explicit `\t` separators matching `csv.DictReader(delimiter='\t')` in train.py

**README.md:**
- Fixed grammar: subject–verb agreement, possessive form, and capitalisation.

**analysis.ipynb:**
- Aligned `language_info.version` with `.python-version` (3.10)

Two items intentionally left as-is: 
1- *fast eval path*: intentional design to run ~3M tokens instead of 20M to stay within resources and examine within the timebudget and user-editable.
2- `PLATFORM["attention"]` branching for the *FlashAttention integration* as it is not within the scope for this PR, which I ran on Turing architecture GPU (RTX 5000).

Looking forward to your insightful comments and review - Thanks again for your time! Shehab @habanwer

File: prepare.py [hunk 1]
```diff
@@ -21,34 +21,142 @@
...
+    _ground = json.load(_f)
+
+_training = _ground["training"]
+_data = _ground["data"]
+_tok = _ground["tokenizer"]
+_mode = _ground["mode"]
+
+MAX_SEQ_LEN = _training["max_seq_len"]
+TIME_BUDGET = _training["time_budget_test"] if _mode == "test" else _training["time_budget_train"]
+EVAL_TOKENS = _training["eval_tokens_multiplier"] * _training["eval_tokens_unit"]
```
[the remark this answers is not in the document]
Comment (habanwer) [reply prepare.py]: Replaced bare `open()` with explicit error handling (`FileNotFoundError`, `JSONDecodeError`) and top-level key validation — comment updated to *reflect `ground.json` is required*, not optional - as per the commit: 
https://github.com/karpathy/autoresearch/commit/61813694e1480eb12dcfa7f4c5eb617eae7c142d

File: README.md [hunk 2]
```diff
@@ -6,6 +6,30 @@
...
+
+### What changed (relative to [karpathy/autoresearch](https://github.com/karpathy/autoresearch) in alignment with the [Adimension](https://theadimension.ch/Introduction.html)'s [DEITY Principles])
+
+**Data.** Hardcoded constants extracted into machine-readable JSON configs. Training platform, data paths, and time budgets live in `ground.json`; architecture and optimization hyperparameters live in `model.json`. Experiment results are logged in three formats: JSON (structured config), TSV (metrics), and Markdown (human/LLM-readable memory).
+
+**Ethics.** Explicit file ownership governance: `ground.json` and `program.md` are user-owned and read-only; `model.json` and `train.py` are agent-owned and editable through the automation experiment. The agent writes experiment memory to `sessions/memory.md` — pinned to the Git branches id. Results are append-only, and a crash handler persists state even on failure. timestamped and SHA-verified outputs are logged.
```
[the remark this answers is not in the document]
Comment (habanwer) [reply README.md]: README.md was fixed: - Fixed grammar: subject–verb agreement, possessive form, and capitalisation.
https://github.com/karpathy/autoresearch/commit/61813694e1480eb12dcfa7f4c5eb617eae7c142d

File: train.py [hunk 3]
```diff
@@ -90,8 +184,26 @@ def forward(self, x, ve, cos_sin, window_size):
...
+        if win >= T:
+            # Full context — use efficient is_causal path
+            y = F.scaled_dot_product_attention(q, k, v, is_causal=True)
+        else:
+            # Sliding window causal — use cached mask
+            key = (T, win)
+            if self._cached_mask_key != key:
+                rows = torch.arange(T, device=q.device).unsqueeze(1)
+                cols = torch.arange(T, device=q.device).unsqueeze(0)
+                self._cached_mask = (cols <= rows) & ((rows - cols) < win)
```
[the remark this answers is not in the document]
Comment (habanwer) [reply train.py]: Addressed in first review — verified with PyTorch 2.6.0: for boolean `attn_mask`, `True` = allowed, not masked-out. The existing mask is correct, and a clarifying comment in commit 6181369 - as per PyTorch docs: https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html
